corporate/lib/stripe.py

```diff
@@ -2832,127 +2947,13 @@ def get_plan_renewal_or_end_date(plan: CustomerPlan, event_time: datetime) -> da
-            proration_fraction = (plan_renewal_or_end_date - ledger_entry.event_time) / (
-                billing_period_end - last_renewal
-            )
-            price_args = {
-                "unit_amount": int(plan.price_per_license * proration_fraction + 0.5),
-                "quantity": ledger_entry.licenses - licenses_base,
-            }
-            description = "Additional license ({} - {})".format(
-                ledger_entry.event_time.strftime("%b %-d, %Y"),
-                plan_renewal_or_end_date.strftime("%b %-d, %Y"),
-            )
-
-        if price_args:
-            plan.invoiced_through = ledger_entry
-            plan.invoicing_status = CustomerPlan.INVOICING_STATUS_STARTED
-            plan.save(update_fields=["invoicing_status", "invoiced_through"])
-            stripe.InvoiceItem.create(
-                currency="usd",
-                customer=plan.customer.stripe_customer_id,
-                description=description,
-                discountable=False,
-                period={
-                    "start": datetime_to_timestamp(ledger_entry.event_time),
-                    "end": datetime_to_timestamp(
-                        get_plan_renewal_or_end_date(plan, ledger_entry.event_time)
-                    ),
-                },
-                idempotency_key=get_idempotency_key(ledger_entry),
-                **price_args,
-            )
-            invoice_item_created = True
-        plan.invoiced_through = ledger_entry
-        plan.invoicing_status = CustomerPlan.INVOICING_STATUS_DONE
-        plan.save(update_fields=["invoicing_status", "invoiced_through"])
-        licenses_base = ledger_entry.licenses
-
-    if invoice_item_created:
-        if plan.charge_automatically:
-            collection_method = "charge_automatically"
-            days_until_due = None
-        else:
-            collection_method = "send_invoice"
-            days_until_due = DEFAULT_INVOICE_DAYS_UNTIL_DUE
-        stripe_invoice = stripe.Invoice.create(
-            auto_advance=True,
-            collection_method=collection_method,
-            customer=plan.customer.stripe_customer_id,
-            days_until_due=days_until_due,
-            statement_descriptor=plan.name,
-        )
-        stripe.Invoice.finalize_invoice(stripe_invoice)
-
-    plan.next_invoice_date = next_invoice_date(plan)
-    plan.save(update_fields=["next_invoice_date"])
-
-
 def invoice_plans_as_needed(event_time: Optional[datetime] = None) -> None:
     if event_time is None:  # nocoverage
         event_time = timezone_now()
+    # TODO: Add RemoteRealmBillingSession and RemoteServerBillingSession cases.
     for plan in CustomerPlan.objects.filter(next_invoice_date__lte=event_time):
-        invoice_plan(plan, event_time)
+        if plan.customer.realm is not None:
+            RealmBillingSession(realm=plan.customer.realm).invoice_plan(plan, event_time)
```

Comment:
We could alternatively rename this to `invoice_realm_plans_as_needed` (or something similar) and use exclude in the query to only get plans with realm set ...

```python
CustomerPlan.objects.filter(next_invoice_date__lte=event_time).exclude(customer__realm__isNull=true)
```

Reply:
I think this is probably a fine pattern.